### PR TITLE
Bug 2649 shield study

### LINKED ISSUE #2649: Revive Shield Study addon wrapper code
We need to blow the dust off of the shield study code which wraps the addon in the shield study layer for:

- starting the addon
- selecting variants (the exact variants right now are unclear, so to start just have two variants: control (tiles with monkeypatched telemetry) and activity stream
- uninstalling it after the study is complete
- monkey patching tiles for rich telemetry

The commit that has all the relevant code is here:

https://github.com/mozilla/activity-stream/commit/1258e6c9b292a0e4b22985fb5a543cf797e193b2

So this needs to be rebased onto master, updated, and then tested in the following versions of Firefox: 54, 55.

We also need to make sure that this study can survive an update from 54 to 55, so figure out if there's any way to get 54, install the shield study addon on it, then force it to update to 55 somehow, and make sure the study continues to function as expected.  

The tiles telemetry monkey patching will only work in 54 and not 55, so we need to make sure that if someone has the study addon installed in 54 and the tiles telemetry is monkey patched, and then their browser updates, that the telemetry just quietly goes away but that the study continues to exist and not crash horribly.

@sarracini is available for mentorship on this because she wrote most of this code, but her time is precious so be sparing in your requests.

### PR DESCRIPTION
**Pull request by piatra**
Tested in nightly. Tried testing in beta but fails due to addon signing process:
```
addons.xpi	WARN	Add-on @activity-stream-shield-study is not correctly signed.
```

**UPDATE** Works in [unbranded builds](https://wiki.mozilla.org/Add-ons/Extension_Signing#Unbranded_Builds)

Fixes #2649

### CODE REVIEW

**Comment by piatra** [comment]
In order to test you cannot use a release version of Firefox 54 because changing `xpinstall.whitelist.required` does nothing in a release version.
You need to [install an unbranded version of Firefox 54](https://wiki.mozilla.org/Add-ons/Extension_Signing#Unbranded_Builds) and run shield-cli experiment with that.
I added a pref to enable more logging in case something fails but we can take that out.

File: addon/shield.js [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+  _connectRetried: 0,
+  _reconnectTimeoutID: null,
+  _initializedTiles: false,
+  _doesHaveTestPilot: false,
+  _eligibilityFilters: [],
+  _placesQueries: null,
+  _originalPreloadPref: null,
+
+  setVariant(variant) {
+    this._variant = `shield-study-01-${variant}`;
```
Comment (sarracini) [inline addon/shield.js]: This should be 
```js
`shield-study-02-${variant}`
```

File: addon/shield.js [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+  setUp: Task.async(function*() {
+    this._originalPreloadPref = prefs.get("browser.newtab.preload");
+    prefs.set("browser.newtab.preload", false);
+    PlacesProvider.links.init();
+    const clientID = yield ClientID.getClientID();
+    this._options.clientID = clientID;
+    this._options.shield_variant = this._variant;
+    this._options.telemetry = false;
+    this._options.tp_version = "1.1.7";
```
Comment (sarracini) [inline addon/shield.js]: i *think* this should reflect the version of the addon that is in test pilot right now, so i guess 1.1.12 ?
@emtwo can you remember why we had this and what this was again?

**Comment by sarracini** [comment]
(Not sure if you know this but) normally our review process is:
1. set the label to "Needs Review"
2. tag a reviewer
3. assign the PR to the person you've assigned to be the reviewer
4. if the reviewer has requested changes they will assign you back to the PR with the label "Needs work"
5. once you've done the changes, repeat from step 1

Also remember to smoosh your commits and change the commit message to something more descriptive

File: addon/shield.js (hunk 2 above)
Comment (piatra) [inline addon/shield.js]: Maybe `"version": "1.12.1",` to reflect the number in the package.json?

**Comment by piatra** [comment]
@sarracini Tested multiple times in 54, no problems getting both variants + always get the survey on uninstalls.

File: addon/shield.js (hunk 2 above)
Comment (sarracini) [inline addon/shield.js]: If yes, then yeah 1.13.0 but i want to double check with @emtwo because there's also an addon version so........

**Comment by sarracini** [comment]
@piatra did you check with @ncloudioj to see if the pings were going through?

**Comment by piatra** [comment]
Added some extra people for review because there are some questions to be answered.
@ncloudioj can you just confirm if there are any other pings that need to be checked
@jaredkerim is there any way to know when the experiment will run. regarding: https://github.com/mozilla/activity-stream/pull/2680#issuecomment-308065087

**Comment by ncloudioj** [comment]
Hey @piatra , let's verify following user events are captured and processed properly by both activity stream and tiles variants. You can trigger those events locally in the Top sites area, and make sure they are stored in the database afterwards.
* search
* click
* dismiss
* bookmark

**Comment by piatra** [comment]
There is no `DISMISS` event right? Looking at the code for the `LinkMenu` I see
```
userEvent: "BLOCK",
```
For the dismiss action. Below are the rest.


screen shot 2017-06-14 at 14 49 09
screen shot 2017-06-14 at 14 49 17
screen shot 2017-06-14 at 14 49 41

screen shot 2017-06-14 at 15 00 57

**Comment by ncloudioj** [comment]
@piatra Yes, we call it `BLOCK` in the reporting. Thanks for confirming this!

LGTM! r=nanj

File: SHIELD.md [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+
+User eligibility criteria:
+
+- Must have about:newtab enabled
+- Must have local set to en-US
+- Must not have Test Pilot add on installed
+
+Findings: *links to dashboards go here*
+
+###shield-study-02
```
Comment (jaredkerim) [inline SHIELD.md]: +###shield-study-02 
+Dates of study: June 26 2017 - August xx 2017
+
+Duration: 14 days
+
+Activity Stream addon version: 1.14.x
+
+Metrics measured: Clicks and blocks on top sites, search, sessions, performance related events
+
+Total number of users: 10,000
+
+Total number of users per arm: 5,000
+
+Control group: 'Tiles'
+
+Variant group: 'Activity Stream'
+
+User eligibility criteria:
+
+- Must have about:newtab enabled
+- Must have local set to en-US
+- Must not have Test Pilot add on installed
+
+Findings: *links to dashboards go here*

File: addon/index.js [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+const {when: unload} = require("sdk/system/unload");
+const {Feature} = require("./shield");
+const {Cu} = require("chrome");
+const feature = new Feature();
+
+const studyConfig = {
+  name: "ACTIVITY_STREAM",
+  days: 14,
+  surveyUrls: {
+    "end-of-study": "https://qsurvey.mozilla.com/s3/Shield-Activity-Stream",
```
Comment (jaredkerim) [inline addon/index.js]: Let's make sure this is the correct link to the new survey.

File: addon/shield.js [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+
+  setUp: Task.async(function*() {
+    this._originalPreloadPref = prefs.get("browser.newtab.preload");
+    prefs.set("browser.newtab.preload", false);
+    PlacesProvider.links.init();
+    const clientID = yield ClientID.getClientID();
+    this._options.clientID = clientID;
+    this._options.shield_variant = this._variant;
+    this._options.telemetry = false;
+    this._options.tp_version = "1.13.0";
```
Comment (jaredkerim) [inline addon/shield.js]: This will be 1.14.0

File: dev-prefs.json [hunk 6]
```diff
@@ -2,6 +2,8 @@
   "dom.ipc.processCount": 1,
   "javascript.options.strict": false,
   "xpinstall.signatures.required": false,
+  "extensions.logging.enabled": true,
   "extensions.@activity-streams.sdk.console.logLevel": "info",
-  "extensions.@activity-streams.telemetry.ping.endpoint": "https://onyx_tiles.stage.mozaws.net/v3/links/activity-stream"
+  "extensions.@activity-streams.telemetry.ping.endpoint": "https://onyx_tiles.stage.mozaws.net/v3/links/activity-stream",
+  "extensions.@activity-stream-shield-study.telemetry.ping.endpoint": "https://onyx_tiles.stage.mozaws.net/v3/links/activity-stream"
```
Comment (jaredkerim) [inline dev-prefs.json]: @ncloudioj Is this the right endpoint to send the shield study telemetry to?

Comment (jaredkerim) [inline dev-prefs.json]: @ncloudioj Oh I see this is dev prefs, for the production one do we send it to the regular production activity stream telemetry endpoint and then separate the shield data using an infernyx job?

Comment (jaredkerim) [inline dev-prefs.json]: In which case we don't want a separate pref I think?

Comment (ncloudioj) [inline dev-prefs.json]: Yes, it shares the same endpoint with the test pilot addon. Each payload of Shield has a "shield_variant" field to identify itself as a shield specific ping. So, we don't need a separate pref here for the endpoint.

**Comment by piatra** [comment]
@jaredkerim 
screen shot 2017-06-16 at 14 17 46
Session events are coming in from all versions.

Click and block events that depend on `DirectoryLinksProvider` do not work in FFx >= 55. 
Using sdk/page-mod you can [listen for click events and it's not that difficult](https://gist.github.com/piatra/15e0b3e145ebc4a147fa679a15281d09) and blocking should be similar or even less code.

I am not sure of the reason but I'm not seeing any events from the past 2 days showing up in staging. But I know it works because I previously provided [screenshots testing for them](https://github.com/mozilla/activity-stream/pull/2680#issuecomment-308424200).
screen shot 2017-06-16 at 14 23 03

File: addon/index.js (hunk 4 above)
Comment (jaredkerim) [inline addon/index.js]: Correct survey link:

https://qsurvey.mozilla.com/s3/Shield-Activity-Stream-2

**Comment by jaredkerim** [comment]
I am closing this PR and moving it to https://github.com/mozilla/activity-stream/pull/2747

**Pull request closed without merging**